### PR TITLE
NoC Turn Model Routing Algorithms

### PR DESCRIPTION
**Pull request by soheilshahrouz**
#### Description
This PR adds four new NoC routing algrotihms from Turn Model paper. XY routing algorithm can only generate a single route for a source/destination location pair. These new algorithms can exploit path diversity in a mesh topology and generate mulitple routes. This PR also adds a new class to check deadlock freedom in generated NoC routes. 

#### Motivation and Context
After adding NoC congestion modeling to VPR, it turned out XY routing algorithm is not able to get rid of congestion in some benchmarks. Turn Model routing algorithms exploit path diversity and distribute traffic flows more evenly across NoC links. When combined with NoC congestion modeling, these algorithm are more effective in reducing NoC link congestion.

The main reason we do not route traffic flows arbitrarily is to avoid deadlock. Turn model routing algorithms guarantee deadlock freedom. However, deadlock freedom is not currently checked in VPR. This PR implements a class to represent channel dependency graph, which is used to detect possible deadlocks in a NoC routing solution.

Traffic flow files for old synthetic benchmarks is updated in this PR. In some old benchmarks, a congestion-free solution was not possible because some traffic flows had a bandwidth higher than the link bandwidth. This PR updates traffic flow bandwidths in flow files so that a congestion-free solution is feasible.

New synthetic benchmarks have been added to represent designs that access HBM channels. In these benchmarks, some NoC routers are locked down at the boundary of device to imitate an HBM channel accessible through a NoC router.

#### How Has This Been Tested?
QoR has been measured on modified synthetic NoC benchmarks.

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by soheilshahrouz** [comment]
[This](https://docs.google.com/spreadsheets/d/1mb8VTgzPTw4_wMXyQs_jnQEfImilRhk7eUnSntKu4qk/edit?usp=sharing) is the link to QoR spreadsheet for synthetic benchmarks. It is still being updated with new results. When completed, I'll post the summary here.

**Comment by vaughnbetz** [comment]
Looks good.
Should compare to the master branch on Titan with all this code, and on the NoC benchmarks.

**Comment by vaughnbetz** [comment]
Also nice to give a normalized / summary tab.

File: vpr/src/noc/channel_dependency_graph.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     *
+     * @param n_links The total number of NoC links.
+     * @param traffic_flow_routes Generated traffic flow routes by the routing
+     * algorithm.
+     */
+    ChannelDependencyGraph(size_t n_links,
+                           const vtr::vector<NocTrafficFlowId, std::vector<NocLinkId>>& traffic_flow_routes);
+
+    /**
+     * @brief Checks whether CDG has any cycles.
```
Comment (vaughnbetz) [inline vpr/src/noc/channel_dependency_graph.h:43]: I'd add "; a cycle implies deadlocks are possible."

File: vpr/src/noc/channel_dependency_graph.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+class ChannelDependencyGraph {
+  public:
+    ChannelDependencyGraph() = delete;
+
+    /**
+     * @brief Constructor
+     *
+     * @param n_links The total number of NoC links.
+     * @param traffic_flow_routes Generated traffic flow routes by the routing
```
Comment (vaughnbetz) [inline vpr/src/noc/channel_dependency_graph.h:36]: "The route of each traffic flow (which could have been generated by any routing algorithm)."

File: vpr/src/noc/channel_dependency_graph.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    // indicates whether a node (NoC link) in CDG is visited during DFS
+    vtr::vector<NocLinkId, bool> visited(n_vertices, false);
+    // indicates whether a node (NoC links) is currently in stack
+    vtr::vector<NocLinkId, bool> on_stack(n_vertices, false);
+    // the stack used to perform graph traversal (DFS). Contains to-be-visited vertices
+    std::stack<NocLinkId> stack;
+
+    // iterate over all vertices (NoC links)
+    for (NocLinkId noc_link_id : adjacency_list_.keys()) {
```
Comment (vaughnbetz) [inline vpr/src/noc/channel_dependency_graph.cpp:60]: Complex way to iterate over an index ....
Could ask Alex and others in the group meeting if you like ...

File: vpr/src/noc/turn_model_routing.h [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+/**
+ * @file
+ * @brief This file declares the TurnModelRouting class, which abstract all
+ * Turn Model routing algorithms. Classes implementing specific Turn Model
+ * algorithms are expected to inherit from this class and comply with its
+ * interface.
+ *
+ * Overview
+ * ========
+ * The TurnModelRouting class abstract Turn Model routing algorithms.
```
Comment (vaughnbetz) [inline vpr/src/noc/turn_model_routing.h:13]: abstracts

File: vpr/src/noc/turn_model_routing.h [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+ * The main idea in Turn Model algorithm is to forbid specific turns
+ * for traffic flows based on the source, destination, and current NoC
+ * router locations in a mesh or torus topology. TurnModelRouting class
+ * exposes a shared interface for all Turn Model routing algorithms.
+ * Derived class can implement specific routing algorithms by implementing
```
Comment (vaughnbetz) [inline vpr/src/noc/turn_model_routing.h:18]: classes

File: vpr/src/noc/turn_model_routing.h [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+ * their override of the exposed interface. More specifically,
+ * get_legal_directions() method returns legal directions that a
```
Comment (vaughnbetz) [inline vpr/src/noc/turn_model_routing.h:20]: the get_...
define the current router.
"on where the source ..."

File: vpr/src/noc/turn_model_routing.h [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+ * traffic flow can follow based on the source, destination, and current
+ * NoC routers are located. select_next_direction() selects one of these
+ * legal directions. TurnModelRouting() method does not implement these
+ * methods, but calls them in route_flow(). For example, XYRouting can be
+ * implemented by override these two methods. get_legal_directions()
```
Comment (vaughnbetz) [inline vpr/src/noc/turn_model_routing.h:25]: overriding

File: vpr/src/noc/turn_model_routing.h [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+  private:
+    /**
+     * @brief Given the direction to travel next, this function determines
+     * the outgoing link that should be used to travel in the intended direction.
+     * Each router may have any number of outgoing links and each link is not
+     * guaranteed to point in the intended direction, So this function makes
+     * sure that the link chosen points in the intended direction.
+     *
+     * @param curr_router_id The physical router on the FPGA that the routing
```
Comment (vaughnbetz) [inline vpr/src/noc/turn_model_routing.h:144]: Updated in this routine and returned by reference.

File: vpr/src/noc/turn_model_routing.h [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+                                  const NocStorage& noc_model);
+
+
+    /**
+     * @brief Computes MurmurHash3 for an array of 32-bit words initialized
+     * with seed.
+     * @param key Contains elements to be hashed
+     * @param seed The initialization value.
+     * @return uint32_t Computed hash value.
+     */
```
Comment (vaughnbetz) [inline vpr/src/noc/turn_model_routing.h:171]: Briefly explain why this is used/helpful in turn routing. We use it with xxx to do yyy.

File: vpr/src/noc/turn_model_routing.h [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+                                                              const NocStorage& noc_model)
+        = 0;
+
+  protected:
+    // get_legal_directions() return a reference to this vector to avoid allocating a new vector
+    // each time it is called
+    std::vector<TurnModelRouting::Direction> returned_legal_direction{4};
+
+  private:
+    std::vector<uint32_t> inputs_to_murmur3_hahser{4};
```
Comment (vaughnbetz) [inline vpr/src/noc/turn_model_routing.h:222]: hahser -> hasher

**Review by vaughnbetz — COMMENTED** [review]
Partly done ... let's edit more tomorrow or Wed.

**Comment by soheilshahrouz** [comment]
Aggregate bandwidth compared to master branch without congestion modeling

| Congestion Weighting Factor | 0.0 | 0.25 | 0.5 | 1.0 | 2.0 | 5.0 |
| --------------- | --------------- | --------------- | --------------- | --------------- | --------------- | --------------- |
| XY    | 1.00 | 1.04 | 1.04 | 1.04 | 1.06 | 1.08 | 
| West First    | 1.00 | 1.08 | 1.08 | 1.10 | 1.11 | 1.12 |
| Odd Even    |  1.00 | 1.04 | 1.06 | 1.07 | 1.08 | 1.09 |
| North Last   | 1.00 | 1.07 | 1.09 | 1.11 | 1.12 | 1.14 |
| Negative First | 1.00 | 1.07 | 1.10 | 1.11 | 1.12 | 1.13 |1.00E+00

In all algorithms, when the congestion weighting grows, the aggregate bandwith increases. To avoid congestion, the placement engine places some routers farer from each other to expose more path diversity, leading to more links being traversed by certain traffic flows, thus resulting in higher overall bandwidth.

When the congestion weighting factor is zero, all algorithms have the same aggregate bandwidth. This is because they all employ minimal routing, generate routes with the same number of links. Since other NoC-related cost terms only depend on the total number of traversed links by each traffic flow, regardless of which links form the route, all algorithms behave the same as XY routing in master without any congestion modeling.

**Comment by soheilshahrouz** [comment]
Aggregate latency compated to master branch with no congestion modeling

| Congestion Weighting Factor | 0.0 | 0.25 | 0.5 | 1.0 | 2.0 | 5.0 |
| --------------- | --------------- | --------------- | --------------- | --------------- | --------------- | --------------- |
| XY    | 1.00 | 1.01 | 1.02 | 1.02 | 1.03 | 1.04 |
| West First    | 1.00 | 1.07 | 1.07 | 1.08 | 1.08 | 1.10 |
| Odd Even    | 1.00 | 1.03 | 1.05 |  1.06 |  1.07 |  1.07 |
| North Last   | 1.00 | 1.04 | 1.06 | 1.07 | 1.09 | 1.10 |
| Negative First | 1.00 | 1.06 | 1.08 | 1.09 | 1.10 | 1.11 |

The aggregate latency shows a similar behavior as the aggregate bandwith. Therefore, the same arguments can be repeated about the aggregate latency.

**Comment by soheilshahrouz** [comment]
Normalized average congestion ratio 

| Congestion Weighting Factor | 0.0 | 0.25 | 0.5 | 1.0 | 2.0 | 5.0 |
| --------------- | --------------- | --------------- | --------------- | --------------- | --------------- | --------------- |
| XY    | 1.00 | 0.35 | 0.35 | 0.35 | 0.36 | 0.34 |
| West First    | 0.88 | 0.11 | 0.10 | 0.08 | 0.08 | 0.06 |
| Odd Even    | 0.79 | 0.10 | 0.11 | 0.10 | 0.08 | 0.06 |
| North Last   | 0.83 | 0.16 | 0.16 | 0.13 | 0.11 | 0.13 |
| Negative First | 0.84 | 0.10 | 0.09 | 0.08 | 0.08 | 0.07 |

Even when congestion ratio is zero, Turn Model algorithms have lower congestion ratio. Even though these algortihms have the same placement as XY algorithm, they can exploit path diversity to some extent and reduce congestion compared to XY algorithm that cannot benefit from path diversity at all.

Setting congestion weighting factor to a non-zero value helps the placement algorithm to reduce the congestion in all algorithms. Increasing the congestion weighting factor slightly improves congestion. However, it should be noted that this comes at the cost of higher aggregate bandwidth and latency.

Odd-even algorithm is performing better than other algorithms. It is abled to reduce congestion by 94% while increasing the aggregate bandwidth only by 9%. Some other algorithms can achieve the same congestion improvement but increase the aggregate bandwith by at least 12%.

**Comment by soheilshahrouz** [comment]
Placement time compared to master without congestion

| Congestion Weighting Factor | 0.0 | 0.25 | 0.5 | 1.0 | 2.0 | 5.0 |
| --------------- | --------------- | --------------- | --------------- | --------------- | --------------- | --------------- |
| XY    | 1.06 |  1.07 |  1.10 |  1.10 |  1.09 | 1.09 |
| West First    | 1.12 | 1.11 | 1.11 | 1.11 | 1.15 | 1.18 |
| Odd Even    | 1.16 | 1.17 |  1.18 |  1.16 |  1.17 |  1.16 
| North Last   | 1.15 | 1.13 | 1.14 | 1.14 | 1.16 | 1.16 |
| Negative First | 1.15 | 1.16 | 1.16 | 1.14 | 1.15 | 1.16 |

Congestion modeling and virtual method overrides in new routing algorithms are the main reasons of runtime increase. The number of swaps during the plecement stays almost constant.

**Comment by vaughnbetz** [comment]
Thanks, these results look good.

File: vpr/src/noc/negative_first_routing.h [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ * the negative-first routing algorithm.
+ *
+ * Overview
+ * ========
+ * The NegativeFirstRouting class performs packet routing between routers in the
+ * NoC using negative-first routing algorithm. In this algorithm, moving towards
+ * negative directions (west and south) has a higher priority. Once the traffic flow
+ * moves towards north or east, it is no longer allowed to take a turn that aligns
+ * the route towards south and west.
+ */
```
Comment (vaughnbetz) [inline vpr/src/noc/negative_first_routing.h:16]: To increase path diversity, our implementation also ... (describe probability of going N/S vs. E/W that you use, and why).

File: vpr/src/noc/odd_even_routing.h [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+                                                                         const NocStorage& noc_model) override;
+
+    TurnModelRouting::Direction select_next_direction(const std::vector<TurnModelRouting::Direction>& legal_directions,
+                                                      NocRouterId src_router_id,
+                                                      NocRouterId dst_router_id,
+                                                      NocRouterId curr_router_id,
+                                                      NocTrafficFlowId traffic_flow_id,
+                                                      const NocStorage& noc_model) override;
+
+    static inline bool is_odd(int number);
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.h:23]: Comment these two (return true if odd).

File: vpr/src/noc/odd_even_routing.h [hunk 13]
```diff
@@ -0,0 +1,27 @@
+#ifndef VTR_ODD_EVEN_ROUTING_H
+#define VTR_ODD_EVEN_ROUTING_H
+
+#include "turn_model_routing.h"
+
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.h:5]: Probably comment what is this class for ... what is the odd/even routing algorithm.

File: vpr/src/noc/odd_even_routing.cpp [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+                returned_legal_direction.push_back(TurnModelRouting::Direction::RIGHT);
+            } else {
+                if (is_odd(compressed_curr_loc.x) || compressed_curr_loc.x == compressed_src_loc.x) {
+                    if (diff_y > 0) {
+                        returned_legal_direction.push_back(TurnModelRouting::Direction::UP);
+                    } else {
+                        returned_legal_direction.push_back(TurnModelRouting::Direction::DOWN);
+                    }
+                }
+                // the destination column is odd and there are more than 1 column left to destination
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.cpp:71]: Should we have some check and action is the column is even?

File: vpr/src/noc/odd_even_routing.cpp [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        } else { // westbound message
+            returned_legal_direction.push_back(TurnModelRouting::Direction::LEFT);
+            if (is_even(compressed_curr_loc.x)) {
+                if (diff_y > 0) {
+                    returned_legal_direction.push_back(TurnModelRouting::Direction::UP);
+                } else {
+                    returned_legal_direction.push_back(TurnModelRouting::Direction::DOWN);
+                }
+            }
```
Comment (vaughnbetz) [inline vpr/src/noc/odd_even_routing.cpp:84]: Should we have a case for is_odd?

File: vpr/src/noc/odd_even_routing.cpp (hunk 14 above)
Comment (vaughnbetz) [reply vpr/src/noc/odd_even_routing.cpp:71]: If it is somehow covered then I think it merits a comment.

File: vpr/src/place/initial_noc_placement.cpp [hunk 16]
```diff
@@ -275,4 +276,12 @@ void initial_noc_placement(const t_noc_opts& noc_opts, int seed) {
...
     // Run the simulated annealing optimizer for NoC routers
     noc_routers_anneal(noc_opts);
+
+    // check if there is any cycles
+    bool has_cycle = noc_routing_has_cycle();
+    if (has_cycle) {
+        VPR_FATAL_ERROR(VPR_ERROR_PLACE,
+                        "At least one cycle was found in NoC channel dependency graph. This may cause a deadlock "
+                        "when packets wait on each other in a cycle.\n");
+    }
```
Comment (vaughnbetz) [inline vpr/src/place/initial_noc_placement.cpp:286]: I think you also check this in check_place now ... would be a good idea to add if you don't.

File: vpr/test/test_xy_routing.cpp [hunk 17]
```diff
@@ -3,15 +3,13 @@
 
 #include "xy_routing.h"
```
Comment (vaughnbetz) [inline vpr/test/test_xy_routing.cpp:5]: Probably good to put anonymous namespace back in. (Good to do for all these test files that had them .... Srivatsan did a nice job on that ).

File: vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_16_noc_2way_ring.flows [hunk 18]
```diff
@@ -0,0 +1,35 @@
+<traffic_flows>
+
```
Comment (vaughnbetz) [inline vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_16_noc_2way_ring.flows:2]: Document what these benchmarks do / how they work / what they can test either in comments in the flow file or in a README that describes a bunch.

File: vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_32_noc_clique.flows [hunk 19]
```diff
@@ -0,0 +1,994 @@
+<traffic_flows>
+    <single_flow src=".*noc_router_adapter_block_1[^\d].*" dst=".*noc_router_adapter_block_2[^\d].*" bandwidth="3.26e4" latency_cons="21e-9" />
```
Comment (vaughnbetz) [inline vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_32_noc_clique.flows:2]: Comment what the purpose / style of this benchmark is.

File: vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_64_noc_bucket_sort.fixed [hunk 20]
```diff
@@ -0,0 +1,17 @@
+noc_router_adapter_block:noc_router_adapter_block_8|slave_tready~reg0	4	157	0	0
```
Comment (vaughnbetz) [inline vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_64_noc_bucket_sort.fixed:1]: Comment that this emulates HBM-style traffic.

File: vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_64_noc_gaussian_elimination.fixed [hunk 21]
```diff
@@ -0,0 +1,5 @@
+noc_router_adapter_block:noc_router_adapter_block_64|slave_tready~reg0	188	4	0	0
```
Comment (vaughnbetz) [inline vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_64_noc_gaussian_elimination.fixed:1]: Comment on what this does, how it emulates gaussian elimination traffic flow ...

File: vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_64_noc_genome_seq.fixed [hunk 22]
```diff
@@ -0,0 +1,2 @@
+noc_router_adapter_block:noc_router_adapter_block_1|slave_tready~reg0	119 4	0	0
```
Comment (vaughnbetz) [inline vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_64_noc_genome_seq.fixed:1]: Comment this.

File: vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_64_noc_page_rank.fixed [hunk 23]
```diff
@@ -0,0 +1,10 @@
+noc_router_adapter_block:noc_router_adapter_block_1|slave_tready~reg0	96	4	0	0
```
Comment (vaughnbetz) [inline vtr_flow/benchmarks/noc/Synthetic_Designs/congestion_traffic_flow_files/complex_64_noc_page_rank.fixed:1]: Comment.

**Review by vaughnbetz — COMMENTED** [review]
Looks good ... some commenting suggestion and a few namespace suggestions attached.

**Comment by soheilshahrouz** [comment]
@vaughnbetz 
I applied your comments. The modifier code is ready for review. Please take a look into it when you have time.

**Comment by soheilshahrouz** [comment]
I added a new sheet to compare the QoR for titan benchmarks on this branch with master. Here is the summary:

| Branch | pack_time | place_time | place_wl | palce_cpd | place_mem |
| --------------- | --------------- | --------------- | --------------- | --------------- | --------------- |
| master        | 522.97 | 1182.15 | 2567328.32 | 16.16 | 5249.09 |
| this branch | 516.90 | 1149.20 | 2567328.32 | 16.16 | 5251.66 |
| ratio            | 0.988  | 0.972     | 1                  | 1        | 1.00049 |

**Pull request merged**